jbpm-flow/src/main/java/org/jbpm/process/core/impl/ObjectCloner.java

```diff
@@ -0,0 +1,479 @@
+
+    }
+    
+    private static <T> Optional<Constructor<?>> getCollectionConstructor (Class<?> collectionClass, Class<?>... paramTypes) {
+        if (unmutableClasses.contains(collectionClass)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.of(collectionClass.getConstructor(paramTypes));
+        } catch (NoSuchMethodException ex) {
+            logger.debug("Copy constructor not found for type {}", collectionClass, ex);
+            unmutableClasses.add(collectionClass);
+            return Optional.empty();
+        }
+    }
+
+    private static Object cloneObject(Object object) {
+        Class<?> objectClass = object.getClass();
+
+        if (object instanceof Cloneable) {
+            try {
+                return objectClass.getMethod("clone").invoke(object);
+            } catch (ReflectiveOperationException ex) {
+                logger.warn("Failing to call clone in a cloneable object!!", ex);
+                return object;
+            }
+        } else {
+            return cloneWithConstructor(object, objectClass);
+        }
+    }
+
+    private static Object cloneWithConstructor(Object object, Class<?> objectClass) {
+        return constructorExecutorMap.computeIfAbsent(objectClass, ObjectCloner::getConstructorExecutor).apply(object);
+    }
+
+    private static UnaryOperator<Object> getConstructorExecutor(Class<?> objectClass) {
+        ConstructorInfo constructorInfo = new ConstructorInfo(objectClass);
+        if (constructorInfo.getCopyConstructor() != null) {
+            return new CopyConstructorExecutor(constructorInfo.getCopyConstructor());
+        } else {
+            MethodInfo methodInfo = new MethodInfo(objectClass);
+            if (inmutable(methodInfo)) {
+                return identity;
+            }
+            MatcherConstructorExecutor result = getConstructorExecutor(constructorInfo, methodInfo);
+            if (result == null) {
+                logger.warn("No suitable constructor found for class {}", objectClass);
+                return identity;
+            }
+            return result;
+        }
+    }
+
+    private static boolean inmutable(MethodInfo classInfo) {
+        return classInfo.getMatchingMethods().isEmpty() && classInfo.getFields().isEmpty();
+    }
+
+    private static MatcherConstructorExecutor getConstructorExecutor(ConstructorInfo constructorInfo, MethodInfo methodInfo) {
+        for (Constructor<?> constructor : constructorInfo.getOtherConstructors()) {
+            Class<?>[] parameters = constructor.getParameterTypes();
+            Collection<ArgResolver> args = new ArrayList<>();
+            boolean found = true;
+            for (int i = 0; found && i < parameters.length; i++) {
+                ArgResolver arg = getArg(parameters[i], methodInfo);
+                found = arg != null;
+                if (found) {
+                    args.add(arg);
+                }
+            }
+            if (found) {
```

Comment:
if the for-loop above is partially executed until one null argument, then found will be "false" but args will contain the previous not null argument, but will return default constructor, is it right?

Reply:
those previusly calculated args are ignored since the next enclosing loop iteration (search for another suitable constructor) is performed.